server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,315 @@
+#include "irods/msi_genquery2.hpp"
+
+#include "irods/irods_at_scope_exit.hpp"
+#include "irods/rs_genquery2.hpp"
+#include "irods/msParam.h"
+#include "irods/irods_re_structs.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_state_table.h"
+#include "irods/rodsError.h"
+#include "irods/rodsErrorTable.h"
+#include "irods/msi_preconditions.hpp"
+#include "irods/irods_exception.hpp"
+
+#include <fmt/format.h>
+#include <boost/any.hpp>
+#include <nlohmann/json.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <iterator>
+#include <map>
+#include <string>
+
+namespace
+{
+    using log_msi = irods::experimental::log::microservice;
+
+    struct genquery2_context
+    {
+        nlohmann::json rows;
+        std::int32_t current_row = -1;
+    }; // struct genquery_context
+
+    // Holds the mappings for each GenQuery2 resultset.
+    std::map<std::size_t, genquery2_context> gq2_context; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+} // anonymous namespace
+
+/// Query the catalog using the GenQuery2 parser.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// The lifetime of a handle and its resultset is tied to the lifetime of the agent.
+/// The value of the handle MUST NOT be viewed as having any meaning. Policy MUST NOT rely on
+/// its structure for any reason as it may change across releases.
```

Comment:
Just for my own understanding, is this due to the API being experimental, or the microservice, or both?

I also noticed that this note is missing from the other microservice documentation but I assume it is no less true of them.

Reply:
That message is for the MSIs only since they deal with handles.

The structure of a handle has no meaning. It should only be viewed as a key or mapping to something within the system. Building systems that rely on the structure of the handle can/will lead to brittle policy because we reserve the right to change the structure of the handle across any release (not just major releases).

Reply:
> I also noticed that this note is missing from the other microservice documentation but I assume it is no less true of them.

Correct. The execute_genquery2 MSI generates handles, while the rest of the MSIs make use of existing handles.